### PR TITLE
ast: grpc-gateway

### PR DESCRIPTION
**Pull request by pysel**
Closes: #288

This PR introduces `grpc-gateway` option as an RPC to `telescope`.

## Tracking
- [x] Implement `createGRPCGatewayMsgClient`
- [x]  Implement `createGRPCGatewayQueryClient`
- [ ] Correctly created request/response types. (future PR, only if required)
- [x] Divide `client/rpc/class`'s rpc types to standalone folders: `tendermint`, `grpc-gateway` and `grpc-web (future PRs)`
- [x] Move general logic from `tendermint`'s `rpc.ts` into standalone `rpc/class/utils` package

## Testing 

Add test for:
- [x] Msg
- [x] Query

## Documentation 
Not currently documented

### CODE REVIEW

**Comment by RusAkh** [comment]
hey @pyramation. I would like to hear your opinion on:

> Divide `client/rpc/class`'s rpc types to standalone folders: `tendermint`, `grpc-gateway` and `grpc-web`.

and 

> Move general logic from `tendermint`'s `rpc.ts` into standalone `rpc/class/utils` package

Do you think we need it?

**Comment by pyramation** [comment]
> > Divide `client/rpc/class`'s rpc types to standalone folders: `tendermint`, `grpc-gateway` and `grpc-web`.
> 
> and
> 
> > Move general logic from `tendermint`'s `rpc.ts` into standalone `rpc/class/utils` package [ref](https://github.com/osmosis-labs/telescope/pull/296/files#diff-b342628aa9d46cd987fae2b8c6749dadb90e90ceffdc5243b18c2143236e3175)
> 
> Do you think we need it?

great! I love that

**Comment by pyramation** [comment]
wow this is looking pretty solid!

File: packages/ast/src/clients/rpc/class/grpc-gateway/utils/grpc-gateway.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    // ex: static Send(request: MsgSend)
+    // paramRequst is an object representing everything in brackets here
+    const paramRequst = identifier(
+        'request',
+        t.tsTypeAnnotation(
+            t.tsTypeReference(
+                t.identifier(requestType),
+            )
+        ),
+        false // todo: work around optional
```
Comment (pyramation) [inline packages/ast/src/clients/rpc/class/grpc-gateway/utils/grpc-gateway.ts:126]: optionality should be similar to https://github.com/osmosis-labs/telescope/blob/db2456e109aa2bbf513c177ccd6cfb689ea49034/packages/ast/src/clients/rpc/class/rpc.ts#L46-L54

Comment (RusAkh) [reply packages/ast/src/clients/rpc/class/grpc-gateway/utils/grpc-gateway.ts:126]: thanks! I just wanted to make sure I do this thing in the future

Comment (RusAkh) [reply packages/ast/src/clients/rpc/class/grpc-gateway/utils/grpc-gateway.ts:126]: overall, what is left until completion:

- build gateway query class (it is different from msg so I can't really use the already build one)
- correctly import all `fm` helper

 and I think basically that's it

**Comment by RusAkh** [comment]
Query client upd:

`grpc-gateway.query.test.ts` currently generates this code:

```
export class Query {
  static Balance(request: QueryBalanceRequest, initRequest?: fm.InitReq): Promise<QueryBalanceResponse> {
    return fm.fetchReq(`/cosmos/bank/v1beta1/balances/${request["address"]}/by_denom?${fm.renderURLSearchParams(request, ["address"])}`, { ...initRequest,
      method: "GET"
    });

  --- snip ---
}
```

File: __fixtures__/outputv2/akash/audit/v1beta1/audit.rpc.msg.ts [hunk 2]
```diff
@@ -1,35 +1,19 @@
 import { Attribute, AttributeSDKType } from "../../base/v1beta1/attribute";
-import { Rpc } from "../../../helpers";
-import * as _m0 from "protobufjs/minimal";
+import * as fm from "../../../grpc-gateway";
 import { MsgSignProviderAttributes, MsgSignProviderAttributesSDKType, MsgSignProviderAttributesResponse, MsgSignProviderAttributesResponseSDKType, MsgDeleteProviderAttributes, MsgDeleteProviderAttributesSDKType, MsgDeleteProviderAttributesResponse, MsgDeleteProviderAttributesResponseSDKType } from "./audit";
-
-/** Msg defines the provider Msg service */
-export interface Msg {
-  /** SignProviderAttributes defines a method that signs provider attributes */
```
Comment (pyramation) [inline __fixtures__/outputv2/akash/audit/v1beta1/audit.rpc.msg.ts:8]: we should look into adding the comments

File: packages/ast/src/clients/rpc/class/grpc-gateway/grpc-gateway.msg.test.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+const store = getTestProtoStore();
+store.traverseAll();
+
+it('GRPC-Gateway Msg Client', () => {
+    const ref = store.findProto('cosmos/bank/v1beta1/tx.proto');
+    const res = traverse(store, ref);
+    const service: ProtoService = getNestedProto(res).Msg;
+    const context = new GenericParseContext(ref, store, store.options);
+    printCode(createGRPCGatewayMsgClass(context, service))
```
Comment (pyramation) [inline packages/ast/src/clients/rpc/class/grpc-gateway/grpc-gateway.msg.test.ts:15]: we should `expectCode` for snapshots

File: packages/ast/src/clients/rpc/class/tendermint/rpc.msg.test.ts [hunk 4]
```diff
@@ -11,7 +11,7 @@ it('RPC Msg Client', () => {
     const res = traverse(store, ref);
     const service: ProtoService = getNestedProto(res).Msg;
     const context = new GenericParseContext(ref, store, defaultTelescopeOptions);
-    expectCode(createRpcClientInterface(context, service))
-    expectCode(createRpcClientClass(context, service))
-    expectCode(createRpcInterface(context, service))
+    printCode(createRpcClientInterface(context, service))
```
Comment (pyramation) [inline packages/ast/src/clients/rpc/class/tendermint/rpc.msg.test.ts:14]: let's use `expectCode` here too

**Pull request merged**